cpp/src/arrow/array/dict_internal.cc

```diff
@@ -44,119 +44,114 @@ using internal::CopyBitmap;
 // ----------------------------------------------------------------------
 // DictionaryType unification
 
-struct UnifyDictionaryValues {
-  MemoryPool* pool_;
-  std::shared_ptr<DataType> value_type_;
-  const std::vector<const DictionaryType*>& types_;
-  const std::vector<const Array*>& dictionaries_;
-  std::shared_ptr<Array>* out_values_;
-  std::vector<std::vector<int32_t>>* out_transpose_maps_;
-
-  template <typename T>
-  enable_if_no_memoize<T, Status> Visit(const T&) {
-    // Default implementation for non-dictionary-supported datatypes
-    return Status::NotImplemented("Unification of ", value_type_,
-                                  " dictionaries is not implemented");
+template <typename T>
+class DictionaryUnifierImpl : public DictionaryUnifier {
+ public:
+  using ArrayType = typename TypeTraits<T>::ArrayType;
+  using DictTraits = typename internal::DictionaryTraits<T>;
+  using MemoTableType = typename DictTraits::MemoTableType;
+
+  DictionaryUnifierImpl(MemoryPool* pool, std::shared_ptr<DataType> value_type)
+      : pool_(pool), value_type_(value_type), memo_table_(pool) {}
+
+  Status Unify(const Array& dictionary) override {
+    const ArrayType& values = checked_cast<const ArrayType&>(dictionary);
+    if (dictionary.null_count() > 0) {
+      return Status::Invalid("Cannot yet unify dictionaries with nulls");
+    }
+    if (!dictionary.type()->Equals(*value_type_)) {
+      return Status::Invalid("Dictionary type different from unifier: ",
+                             dictionary.type()->ToString());
+    }
+    for (int64_t i = 0; i < values.length(); ++i) {
+      memo_table_.GetOrInsert(values.GetView(i));
+    }
+    return Status::OK();
   }
 
-  template <typename T>
-  enable_if_memoize<T, Status> Visit(const T&) {
-    using ArrayType = typename TypeTraits<T>::ArrayType;
-    using DictTraits = typename internal::DictionaryTraits<T>;
-    using MemoTableType = typename DictTraits::MemoTableType;
-
-    MemoTableType memo_table(pool_);
-    if (out_transpose_maps_ != nullptr) {
-      out_transpose_maps_->clear();
-      out_transpose_maps_->reserve(types_.size());
+  Status Unify(const Array& dictionary, std::shared_ptr<Buffer>* out) override {
+    const ArrayType& values = checked_cast<const ArrayType&>(dictionary);
+    if (dictionary.null_count() > 0) {
+      return Status::Invalid("Cannot yet unify dictionaries with nulls");
```

Comment:
Do we have a JIRA open to add null support to MemoTables?

Reply:
@fsaintjacques ?

Reply:
Nulls are supported in MemoTables (GetOrInsertNull). The issue to adding them in dictionary is compat with parquet/pandas testing facility. There was also deciding if it could have multiple nulls in the dictionary.

Reply:
I'll open a follow up

https://issues.apache.org/jira/browse/ARROW-6628